### PR TITLE
KAFKA-13769 Add tests for ForeignJoinSubscriptionProcessorSupplier

### PR DESCRIPTION
**Pull request by Gerrrr**
This PR introduces a test suite that could've caught the issue hotfixed in https://github.com/apache/kafka/pull/12420.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+import org.apache.kafka.streams.processor.api.MockProcessorContext;
+import org.apache.kafka.streams.processor.api.MockProcessorContext.CapturedForward;
+import org.apache.kafka.streams.processor.api.Processor;
+import org.apache.kafka.streams.processor.api.ProcessorContext;
+import org.apache.kafka.streams.processor.api.Record;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ForeignJoinSubscriptionProcessorSupplierTest {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java:35]: Thanks for this.

I want to make sure we don't overlook this again if we add another version. I think it should be good enough if you add another test that asserts that `SubscriptionWrapper.CURRENT_VERSION == VERSION_1`, along with a comment explaining that when we bump the version, we should also add a new set of tests. WDYT?

**Review by vvcephei — APPROVED** [review]
Thanks, @Gerrrr !

I had a small suggestion to help us keep the new test up-to-date in case we bump the version in the future. Otherwise, the change looks good to me!

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java (hunk 1 above)
Comment (Gerrrr) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java:35]: Good idea! Added:

https://github.com/apache/kafka/blob/b44b119a2c4e89da24b6fe3580771e9acf03a67f/streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java#L46-L50

**Comment by vvcephei** [comment]
These failures are certainly unrelated, but I'm going to re-run the tests before merging, since we didn't get a single green build this time.

```
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.clients.consumer.KafkaConsumerTest.testPollReturnsRecords()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/3/testReport/junit/org.apache.kafka.clients.consumer/KafkaConsumerTest/Build___JDK_8_and_Scala_2_12___testPollReturnsRecords__/)
    [Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testKeyStoreAlter()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/3/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_11_and_Scala_2_13___testKeyStoreAlter__/)
    [Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testKeyStoreAlter()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/3/testReport/junit/kafka.server/DynamicBrokerReconfigurationTest/Build___JDK_11_and_Scala_2_13___testKeyStoreAlter___2/)
    [Build / JDK 17 and Scala 2.13 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/3/testReport/junit/kafka.network/ConnectionQuotasTest/Build___JDK_17_and_Scala_2_13___testListenerConnectionRateLimitWhenActualRateAboveLimit__/)
```

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java (hunk 1 above)
Comment (bellemare) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/ForeignJoinSubscriptionProcessorSupplierTest.java:35]: Smart addition :)

**Review by bellemare — APPROVED** [review]
LGTM! Thanks for this important set of tests

**Comment by vvcephei** [comment]
Ok, at least we got on passing build that time. The failures are again unrelated:
```
[Build / PowerPC / Run PowerPC Build / org.apache.kafka.common.security.oauthbearer.secured.RefreshingHttpsJwksTest.testBasicScheduleRefresh()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/4/testReport/junit/org.apache.kafka.common.security.oauthbearer.secured/RefreshingHttpsJwksTest/Build___PowerPC___Run_PowerPC_Build___testBasicScheduleRefresh__/)
    [Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/4/testReport/junit/org.apache.kafka.connect.integration/ConnectorRestartApiIntegrationTest/Build___JDK_17_and_Scala_2_13___testMultiWorkerRestartOnlyConnector/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest.testReplication()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/4/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationBaseTest/Build___JDK_11_and_Scala_2_13___testReplication__/)
    [Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/4/testReport/junit/org.apache.kafka.connect.integration/ExactlyOnceSourceIntegrationTest/Build___JDK_11_and_Scala_2_13___testSeparateOffsetsTopic/)
    [Build / JDK 11 and Scala 2.13 / kafka.server.RaftClusterSnapshotTest.testSnapshotsGenerated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12437/4/testReport/junit/kafka.server/RaftClusterSnapshotTest/Build___JDK_11_and_Scala_2_13___testSnapshotsGenerated__/)
```

**Pull request merged**